include/realtime_tools/realtime_barrier.hpp

```diff
@@ -0,0 +1,468 @@
+    inline typename std::enable_if_t<std::is_trivially_assignable<realtime, realtimeness>::value,
+      const T *>
+    _get() const
+    {
+      return obj_ ?
+             obj_ :
+             lock_strategy::owns_lock() ?
+             obj_ = mem_->rt_ :
+               nullptr;
+    }
+
+    template<class realtimeness = access_mode>
+    inline typename std::enable_if_t<std::is_trivially_assignable<non_realtime,
+      realtimeness>::value, const T *>
+    _get() const
+    {
+      return obj_ ?
+             obj_ :
+             lock_strategy::owns_lock() ?
+             obj_ = mem_->nrt_ :
+               nullptr;
+    }
+  };
+
+
+  MemoryBarrier()
+  : nrt_(new T()), rt_(new T()), polarity_(false), new_data_available_(false)
+  {
+  }
+
+  explicit MemoryBarrier(const T & data)
+  : nrt_(new T(data)), rt_(new T(data)), polarity_(false), new_data_available_(false)
+  {
+  }
+
+  explicit MemoryBarrier(MemoryBarrier<T> && move) noexcept
+  : nrt_(move.nrt_), rt_(move.rt_), polarity_(move.polarity_), new_data_available_(
+      move.new_data_available_)
+  {
+    move.nrt_ = nullptr;
+    move.rt_ = nullptr;
+    move.polarity_ = false;
+    move.new_data_available_ = false;
+  }
+
+  // allow moving but not copying
+  MemoryBarrier(const MemoryBarrier<T> &) = delete;
+
+  MemoryBarrier & operator=(const MemoryBarrier<T> &) = delete;
+
+
+  // todo: perhaps we need a new_data on both nrt and rt side
+  // true if new data is available, depends on if the memory is being used for Read
+  // from RT or Write to RT mode
+  inline bool new_data_available() const
+  {return new_data_available_;}
+
+  void initialize(const T & value)
+  {
+    wait_lock guard(mutex_);
+    if (guard.owns_lock()) {
+      *nrt_ = *rt_ = value;
+    } else {
+      throw std::runtime_error("request to initialize realtime barrier failed trying to lock");
+    }
+  }
+
+protected:
+  T * nrt_;
+  T * rt_;
+  bool polarity_;  // flipped each time barrier rotates (swaps)
+  bool new_data_available_;
+
+  // Set as mutable so that readFromNonRT() can be performed on a const buffer
+  mutable std::mutex mutex_;
+
+  inline void swap()
+  {
+    // swap pointers
+    T * tmp = nrt_;
+    nrt_ = rt_;
+    rt_ = tmp;
+    polarity_ = !polarity_;
+  }
+
+  friend class DirectAccess<realtime>;
+
+  friend class DirectAccess<non_realtime>;
+};
+
+/// @brief Create a barrier for reading data from a realtime thread
+/// ReadBarrier implements reading data from a realtime thread using a MemoryBarrier. For example,
+/// this is used to transfer state data out of hardware interfaces. The default constructor will
+// create a new memory barrier for use. There is an alternate constructor if you want to create
+// your own MemoryBarrier explicitly.
+template<class T, class locking_strategy = wait_lock>
+class ReadBarrier
+{
+public:
+  using MemoryBarrierType = MemoryBarrier<T>;
+
+  ReadBarrier() noexcept
+  : mem_(new MemoryBarrierType()), owns_mem(true)
+  {
+  }
```

Comment:
I'm not going to repeat this on every function where it could apply so as to avoid spam, but this rule of simple blocks on the same line can be applied to a bunch of functions.